source/simulator/core.cc

```diff
@@ -1697,7 +1697,11 @@ namespace aspect
 
               ++iteration;
             }
-          while (iteration < parameters.max_nonlinear_iterations);
+          while (iteration < parameters.max_nonlinear_iterations
+                 &&
+                 (iteration < parameters.max_nonlinear_iterations_in_prerefinment
+                  ||
+                  parameters.initial_adaptive_refinement <= pre_refinement_step));
```

Comment:
This condition is exceedingly hard to read. It may be correct, but maybe the following would be easier to read?
```
  while ( ! ((iteration >= param.max_n_i) // regular time step
                 ||
                 ((pre_refinement_step <= param.initial_a_r)  // we're in pre-refinement
                  &&
                  (iteration >= ...)))
```
I think this is the same except that the last condition uses `>=` instead of `not <=`!?